addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/RemoteControllerWebSocket.java

```diff
@@ -0,0 +1,678 @@
+            if (isConnecting || isConnected()) {
+                logger.trace("{} already connecting or connected", this.getClass().getSimpleName());
+                return;
+            }
+
+            try {
+                logger.debug("{} connecting to: {}", this.getClass().getSimpleName(), uri);
+
+                isConnecting = true;
+                client.connect(this, uri, new ClientUpgradeRequest());
+            } catch (Exception e) {
+                throw new RemoteControllerException(e);
+            }
+        }
+
+        @Override
+        public void onWebSocketConnect(Session session) {
+            logger.debug("{} connection established: {}", this.getClass().getSimpleName(),
+                    session.getRemoteAddress().getHostString());
+            super.onWebSocketConnect(session);
+
+            isConnecting = false;
+        }
+
+        void close() {
+            getSession().close();
+        }
+
+        void sendCommand(String cmd) {
+            try {
+                // retry openening connection just in case
+                openConnection();
+
+                if (isConnected()) {
+                    getRemote().sendString(cmd);
+                    logger.trace("{}: sendCommand: {}", this.getClass().getSimpleName(), cmd);
+                } else {
+                    logger.warn("{} sending command while socket not connected: {}", this.getClass().getSimpleName(),
+                            cmd);
+                }
+            } catch (Exception e) {
+                logger.error("{}: cannot send command", this.getClass().getSimpleName(), e);
+            }
+        }
+
+        @Override
+        public void onWebSocketText(String str) {
+            logger.trace("{}: onWebSocketText: {}", this.getClass().getSimpleName(), str);
+        }
+
+    }
+
+    // temporary storage for source app. Will be used as value for the sourceApp channel when information is complete
+    private String currentSourceApp = null;
+    // last app in the apps list: used to detect when status information is complete
+    String lastApp = null;
+
+    /**
+     * Retrieve app status for all apps. In the WebSocketv2 handler the currently running app will be determined
+     */
+    private void updateCurrentApp() {
+        if (webSocketV2.isNotConnected()) {
+            logger.warn("Cannot retrieve current app webSocketV2 is not connected");
+            return;
+        }
+
+        currentSourceApp = null;
+        lastApp = null;
+
+        // retrieve last app (don't merge with next loop as this might run asynchronously
+        for (App app : apps.values()) {
+            lastApp = app.appId;
+        }
+
+        for (App app : apps.values()) {
+            getAppStatus(app.appId);
+        }
+    }
+
+    /**
+     * Websocket class for remote control
+     *
+     * @author Arjan Mels
+     *
+     */
+    class WebSocketRemote extends WebSocketBase {

```

Comment:
I suppose you will get at least an informational warning from the SAT that you should mention your contribution for this class, please take a looks at the the report of [the Static code analysis](https://docs.openhab.org/developers/development/bindings.html#static-code-analysis) this might include some new things introduced by you.

Reply:
Done